app/src/main/java/io/mochahub/powermeter/workoutsession/workoutsessiondialog/NewWorkoutDialog.kt

```diff
@@ -0,0 +1,211 @@
+        )
+
+        viewModel =
+            NewWorkoutViewModel(
+                db = AppDatabase(requireContext())
+            )
+        workoutController =
+            WorkoutController(
+                ArrayAdapter(
+                    requireContext(),
+                    R.layout.dropdown_menu_popup_item
+                ), this
+            )
+    }
+
+    override fun onStart() {
+        super.onStart()
+
+        if (dialog != null) {
+            val width = ViewGroup.LayoutParams.MATCH_PARENT
+            val height = ViewGroup.LayoutParams.MATCH_PARENT
+            dialog!!.window!!.setLayout(width, height)
+        }
+    }
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        return inflater.inflate(R.layout.dialog_new_workout, container, false)
+    }
+
+    override fun onDismiss(dialog: DialogInterface) {
+        super.onDismiss(dialog)
+
+        val date = viewModel.simpleDateFormat.parse(newWorkoutDateText.text.toString()).toInstant()
+        val workoutSession = WorkoutSession(newWorkoutNameText.text.toString(), date, workouts)
+        CoroutineScope(Dispatchers.IO).launch {
+            viewModel.saveWorkoutSession(workoutSession)
+        }
+        if (args.workoutSessionID != null) {
+            CoroutineScope(Dispatchers.IO).launch {
+                viewModel.deleteWorkoutSession(args.workoutSessionID!!)
+            }
+        }
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        initFields()
+        initDatePicker()
+
+        newWorkoutToolbar.apply {
+            title = resources.getString(if (args.workoutSessionID == null) R.string.new_workout else R.string.edit_workout)
+            setNavigationOnClickListener { dismiss() }
+        }
+
+        // init with an empty workout to edit
+        if (workouts.isEmpty()) {
+            this.addEmptyWorkout()
+        }
+
+        addWorkoutBtn.setOnClickListener {
+            this.addEmptyWorkout()
+            workoutController.setData(workouts)
+        }
+
+        viewModel.exercises.observe(viewLifecycleOwner, Observer {
+            val adapter = ArrayAdapter<String>(
+                requireContext(), R.layout.dropdown_menu_popup_item,
+                viewModel.exercises.value?.map { it.name } ?: listOf())
+            workoutController.setAdapter(adapter)
+            recyclerView.setController(workoutController)
+            workoutController.setData(workouts)
+        })
+    }
+    // //////////////////////////////////////////////////////////////
+    // Helpers
+    // //////////////////////////////////////////////////////////////
+
+    private fun addEmptyWorkout() {
+        val workout = Workout(
+            exercise = Exercise("", 0.0, ""),
+            sets = ArrayList()
+        )
+        workouts.add(0, workout)
```

Comment:
Why index 0? Seems more natural to me to just do `workouts.add(workout)` and leave it chronological

Reply:
It's for the order. This way the new workouts will be at the top, instead of the bottom. But now that I think about it, it's probably better to have new workouts at the bottom instead, that way its closer to the `Add Set` button.

Reply:
Yeah I think they're better at the bottom

Reply:
My previous comment was mixing up workouts and workoutsets. New workouts are added with the fab, it makes sense for new ones to appear at the top. When you wanna add a new workout, you don't want to have to scroll to the bottom each time you want to edit it.

Reply:
I guess its a stylistic choice, I think it makes more sense for them to populate at the bottom 🤷‍♂

Reply:
Not a blocker though